### PR TITLE
feat: support tencent vector db

### PR DESCRIPTION
**Pull request by quicksandznzn**
# Description

Support Tencent Vector DB

dependencies

tcvectordb==1.3.2

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [ ] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] TODO

# Suggested Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods
- [ ] `optional` I have made corresponding changes to the documentation 
- [ ] `optional` I have added tests that prove my fix is effective or that my feature works
- [ ] `optional` New and existing unit tests pass locally with my changes

### CODE REVIEW

File: api/.env
Comment (crazywoola) [inline api/.env:1]: This should not be committed into the vcs. It should remain ignored.

File: api/tests/unit_tests/core/rag/datasource/vdb/tencent/test_tencent.py
Comment (crazywoola) [inline api/tests/unit_tests/core/rag/datasource/vdb/tencent/test_tencent.py:1]: Tests are needed.

File: docker/docker-compose.yaml [hunk 1]
```diff
@@ -229,6 +229,14 @@ services:
       RELYT_USER: postgres
       RELYT_PASSWORD: difyai123456
       RELYT_DATABASE: postgres
+      # tencent configurations
+      TENCENT_URL: http://127.0.0.1
```
Comment (crazywoola) [inline docker/docker-compose.yaml]:
```suggestion
      TENCENT_VECTOR_DB_URL: http://127.0.0.1
```
To distinguish with other Tencent services. The rest should be the same.

File: api/config.py [hunk 2]
```diff
@@ -228,6 +228,15 @@ def __init__(self):
         self.RELYT_PASSWORD = get_env('RELYT_PASSWORD')
         self.RELYT_DATABASE = get_env('RELYT_DATABASE')
 
+        # tencent settings
+        self.TENCENT_URL = get_env('TENCENT_URL')
```
Comment (crazywoola) [inline api/config.py]: See the suggestions below in the `docker-compose.yaml`.

**Review by crazywoola — CHANGES_REQUESTED** [review]

**Comment by bowenliang123** [comment]
1. tencent vdb , or `Tencent Cloud VectorDB` , is not an open-sourced vector db , which leads to less testability to test against the target vdb instance. The code will easily come to an idle status.
2. missing required `tcvectordb` python package in requirements.txt. 
3. the package `tcvectordb` provides literally no information for usage and requirements
4. never put `.env` file to the PR

File: api/.env
Comment (quicksandznzn) [reply api/.env:1]: resolve

File: api/tests/unit_tests/core/rag/datasource/vdb/tencent/test_tencent.py
Comment (quicksandznzn) [reply api/tests/unit_tests/core/rag/datasource/vdb/tencent/test_tencent.py:1]: resolve

File: docker/docker-compose.yaml (hunk 1 above)
Comment (quicksandznzn) [reply docker/docker-compose.yaml]: resolve

File: api/core/rag/datasource/vdb/tencent/tencent_vector.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+                                                                params=document.HNSWSearchParams(
+                                                                    ef=kwargs.get("ef", 10)),
+                                                                retrieve_vector=False,
+                                                                limit=kwargs.get('top_k', 4),
+                                                                timeout=self._client_config.timeout,
+                                                                )
+        return self._get_search_res(res)
+
+    def search_by_full_text(self, query: str, **kwargs: Any) -> list[Document]:
+        # res = (self._db.collection(self._collection_name)
```
Comment (crazywoola) [inline api/core/rag/datasource/vdb/tencent/tencent_vector.py]: These comments can be removed for now.

**Review by crazywoola — COMMENTED** [review]
The rest looks good to me.

**Comment by bowenliang123** [comment]
It's alright to me, if tests are properly provided and tested on the mocked embedded tencent vdb instance.
LGTM.

File: api/core/rag/datasource/vdb/tencent/tencent_vector.py (hunk 3 above)
Comment (quicksandznzn) [reply api/core/rag/datasource/vdb/tencent/tencent_vector.py]: resolve

**Comment by wade30822** [comment]
how is it going?

**Comment by quicksandznzn** [comment]
> how is it going?

wait review ~

**Comment by bowenliang123** [comment]
1. please resolve the sytle violation in Python code by running `dev/reformat`.
2. move the tests to `api/tests/integration_tests/vdb/tcvectordb`

**Comment by quicksandznzn** [comment]
> 1. please resolve the sytle violation in Python code by running `dev/reformat`.
> 2. move the tests to `api/tests/integration_tests/vdb/tcvectordb`

done~

**Comment by bowenliang123** [comment]
ok, thx.

File: api/core/rag/datasource/vdb/tencent/tencent_vector.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    def search_by_vector(self, query_vector: list[float], **kwargs: Any) -> list[Document]:
+
+        res = self._db.collection(self._collection_name).search(vectors=[query_vector],
+                                                                params=document.HNSWSearchParams(
+                                                                    ef=kwargs.get("ef", 10)),
+                                                                retrieve_vector=False,
+                                                                limit=kwargs.get('top_k', 4),
+                                                                timeout=self._client_config.timeout,
+                                                                )
+        return self._get_search_res(res)
```
Comment (JohnJyong) [inline api/core/rag/datasource/vdb/tencent/tencent_vector.py]: the socre is not returned and we have the score thresold check.
pls refer to below code:

```
        for doc, score in docs_and_scores:
            score_threshold = kwargs.get("score_threshold", .0) if kwargs.get('score_threshold', .0) else 0.0
            # check score threshold
            if score > score_threshold:
                doc.metadata['score'] = score
                docs.append(doc)
```

Comment (JohnJyong) [reply api/core/rag/datasource/vdb/tencent/tencent_vector.py]: @quicksandznzn

**Comment by JohnJyong** [comment]
> @quicksandznzn In the method `def search_by_vector()`, it return the results without score_threshold filtering, and tcvectordb won't return the score. Score_threshold won't work as expect. For example, you want the agent to ask for more information instead of giving an irrelevant reply. It seems a problem of the sdk.



> @quicksandznzn In the method `def search_by_vector()`, it return the results without score_threshold filtering, and tcvectordb won't return the score. Score_threshold won't work as expect. For example, you want the agent to ask for more information instead of giving an irrelevant reply. It seems a problem of the sdk.

the sdk is fine , it has returned the score

File: api/core/rag/datasource/vdb/tencent/tencent_vector.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    def to_index_struct(self) -> dict:
+        return {
+            "type": self.get_type(),
+            "vector_store": {"class_prefix": self._collection_name}
+        }
+
+    def _create_collection(self, dimension: int) -> None:
+        lock_name = 'vector_indexing_lock_{}'.format(self._collection_name)
+        with redis_client.lock(lock_name, timeout=20):
+            self.delete()
```
Comment (JohnJyong) [inline api/core/rag/datasource/vdb/tencent/tencent_vector.py:89]: we need to check the collection is exist or not，if the collection is not exist ， we need to create the collection . if  the collection is exist , just return the collection @quicksandznzn

**Comment by quicksandznzn** [comment]
> ```
>         score_threshold = kwargs.get("score_threshold", .0) if kwargs.get('score_threshold', .0) else 0.0
>         return self._get_search_res(res, score_threshold)
> ```
> 
> ```
>     def _get_search_res(self, res, score_threshold):
>         docs = []
>         if res is None or len(res) == 0:
>             return docs
> 
>         for result in res[0]:
>             meta = result.get(self.field_metadata)
>             if meta is not None:
>                 meta = json.loads(meta)
>             score = 1 - result.get("score")
>             if score > score_threshold:
>                 meta['score'] = score
>                 doc = Document(page_content=result.get(self.field_text), metadata=meta)
>                 docs.append(doc)
>         return docs
> ```

thanks , optimized

**Comment by quicksandznzn** [comment]
> > ```
> >         score_threshold = kwargs.get("score_threshold", .0) if kwargs.get('score_threshold', .0) else 0.0
> >         return self._get_search_res(res, score_threshold)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> >     def _get_search_res(self, res, score_threshold):
> >         docs = []
> >         if res is None or len(res) == 0:
> >             return docs
> > 
> >         for result in res[0]:
> >             meta = result.get(self.field_metadata)
> >             if meta is not None:
> >                 meta = json.loads(meta)
> >             score = 1 - result.get("score")
> >             if score > score_threshold:
> >                 meta['score'] = score
> >                 doc = Document(page_content=result.get(self.field_text), metadata=meta)
> >                 docs.append(doc)
> >         return docs
> > ```
> 
> thanks , optimized

Optimized,Refer to your suggestions

File: api/core/rag/datasource/vdb/tencent/tencent_vector.py (hunk 5 above)
Comment (JohnJyong) [reply api/core/rag/datasource/vdb/tencent/tencent_vector.py:89]: @quicksandznzn 
```
            collections = self._db.list_collections()
            for collection in collections:
                if collection.collection_name == self._collection_name:
                    self.collection = collection
                    return
```

**Comment by crazywoola** [comment]
@quicksandznzn Hello please add my wechat crazyphage, I will invite. you to our contributors' group

**Comment by quicksandznzn** [comment]
> crazyphage

yep

**Comment by JohnJyong** [comment]
This branch has conflicts that must be resolved , pls fix it , thanks @quicksandznzn

File: api/core/rag/datasource/vdb/tencent/tencent_vector.py [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+            if collection.collection_name == self._collection_name:
+                return True
+        return False
+
+    def _create_collection(self, dimension: int) -> None:
+        lock_name = 'vector_indexing_lock_{}'.format(self._collection_name)
+        with redis_client.lock(lock_name, timeout=20):
+            collection_exist_cache_key = 'vector_indexing_{}'.format(self._collection_name)
+            if redis_client.get(collection_exist_cache_key):
+                return
```
Comment (JohnJyong) [inline api/core/rag/datasource/vdb/tencent/tencent_vector.py:84]: the  `self.collection`  not initialized @quicksandznzn

Comment (quicksandznzn) [reply api/core/rag/datasource/vdb/tencent/tencent_vector.py:84]: fix

**Comment by zeroameli** [comment]
@quicksandznzn I found some problems:
https://github.com/langgenius/dify/blob/a591366f727da7d6ae3612aaec678fe081e84e11/api/core/rag/datasource/vdb/tencent/tencent_vector.py#L158-L161
- param `limit` is needed when query with filter, why not use delete with filter.
- fields in metadata should be indexed if we need to filter by them.

The self.collection won't be initialized in multithread because of redis lock (For example, create a dataset), why not use `self._db.collection(self._collection_name)`

**Comment by wade30822** [comment]
it takes so long~~ 😭

**Review by crazywoola — DISMISSED** [review]

**Review by crazywoola — APPROVED** [review]

**Pull request merged**